From 9aa6376e629864cc949807f9e7da3864ef51ea12 Mon Sep 17 00:00:00 2001
From: Wang Hui <Hui.Wang@windriver.com>
Date: Fri, 18 Jan 2013 17:12:01 +0800
Subject: [PATCH 47/50] tty: 8250: add mv_armada port type

If we don't set port type for armadaxp serial, it will fall into a
unknown type which means there is no fifo. But the manual of armadaxp
says there is 16bytes fifo, so we add a port type for armadaxp serial
port which can greatly decrease the number of serial port interrupt.

Signed-off-by: Wang Hui <Hui.Wang@windriver.com>
---
 drivers/tty/serial/8250/8250.c    |    7 +++++++
 drivers/tty/serial/8250/8250_dw.c |    1 +
 include/linux/serial_core.h       |    3 ++-
 3 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/drivers/tty/serial/8250/8250.c b/drivers/tty/serial/8250/8250.c
index dff6214..dd7d86b 100644
--- a/drivers/tty/serial/8250/8250.c
+++ b/drivers/tty/serial/8250/8250.c
@@ -282,6 +282,13 @@ static const struct serial8250_config uart_config[] = {
 		.fcr		= UART_FCR_ENABLE_FIFO | UART_FCR_R_TRIG_10,
 		.flags		= UART_CAP_FIFO | UART_CAP_AFE | UART_CAP_EFR,
 	},
+	[PORT_MV_ARMADA] = {
+		.name		= "Armada",
+		.fifo_size	= 16,
+		.tx_loadsz	= 1,
+		.fcr		= UART_FCR_ENABLE_FIFO | UART_FCR_R_TRIG_10,
+		.flags		= UART_CAP_FIFO,
+	},
 };
 
 /* Uart divisor latch read */
diff --git a/drivers/tty/serial/8250/8250_dw.c b/drivers/tty/serial/8250/8250_dw.c
index 4fecf11..1a87193 100644
--- a/drivers/tty/serial/8250/8250_dw.c
+++ b/drivers/tty/serial/8250/8250_dw.c
@@ -148,6 +148,7 @@ struct plat_serial8250_port *p = pdev->dev.platform_data;
 	}
 	port.uartclk = val;
 #else
+	port.type = PORT_MV_ARMADA;
 	port.iotype = p->iotype;
 	port.serial_in = dw8250_serial_in32;
 	port.serial_out = dw8250_serial_out32;
diff --git a/include/linux/serial_core.h b/include/linux/serial_core.h
index 3b7e2b6..62464f6 100644
--- a/include/linux/serial_core.h
+++ b/include/linux/serial_core.h
@@ -47,7 +47,8 @@
 #define PORT_U6_16550A	19	/* ST-Ericsson U6xxx internal UART */
 #define PORT_TEGRA	20	/* NVIDIA Tegra internal UART */
 #define PORT_XR17D15X	21	/* Exar XR17D15x UART */
-#define PORT_MAX_8250	21	/* max port ID */
+#define PORT_MV_ARMADA	22
+#define PORT_MAX_8250	22	/* max port ID */
 
 /*
  * ARM specific type numbers.  These are not currently guaranteed
-- 
1.7.0

